### PR TITLE
LibJS: Do not revisit already visited values in update_function_name()

### LINKED ISSUE #3471: LibJS: Endless recursion in update_function_name
Found while fuzzing.
File: [crash-endless-recursion.txt](https://github.com/SerenityOS/serenity/files/5213316/crash-endless-recursion.txt)

GDB trace:
```
#0  0x00007ffff7ae4fce in _int_malloc (av=av@entry=0x7ffff7c35b80 <main_arena>, bytes=bytes@entry=696) at malloc.c:3679
#1  0x00007ffff7ae7419 in __GI___libc_malloc (bytes=696) at malloc.c:3066
#2  0x0000000000628227 in AK::Vector<JS::ValueAndAttributes, 0ul>::ensure_capacity (this=0x7fffff7ff138, needed_capacity=29) at .././AK/Vector.h:450
#3  AK::Vector<JS::ValueAndAttributes, 0ul>::grow_capacity (this=0x7fffff7ff138, needed_capacity=20) at .././AK/Vector.h:442
#4  AK::Vector<JS::ValueAndAttributes, 0ul>::append (this=0x7fffff7ff138, value=...) at .././AK/Vector.h:386
#5  JS::IndexedProperties::values_unordered (this=<optimized out>) at ../Libraries/LibJS/Runtime/IndexedProperties.cpp:363
#6  0x00000000005df76b in JS::update_function_name (value=..., name=...) at ../Libraries/LibJS/AST.cpp:63
...
#104732 0x00000000005df79f in JS::update_function_name (value=..., name=...) at ../Libraries/LibJS/AST.cpp:64
#104733 0x00000000005df79f in JS::update_function_name (value=..., name=...) at ../Libraries/LibJS/AST.cpp:64
#104734 0x00000000005df79f in JS::update_function_name (value=..., name=...) at ../Libraries/LibJS/AST.cpp:64
#104735 0x00000000005df79f in JS::update_function_name (value=..., name=...) at ../Libraries/LibJS/AST.cpp:64
#104736 0x00000000005df79f in JS::update_function_name (value=..., name=...) at ../Libraries/LibJS/AST.cpp:64
#104737 0x00000000005df79f in JS::update_function_name (value=..., name=...) at ../Libraries/LibJS/AST.cpp:64
#104738 0x00000000005df79f in JS::update_function_name (value=..., name=...) at ../Libraries/LibJS/AST.cpp:64
#104739 0x00000000005df79f in JS::update_function_name (value=..., name=...) at ../Libraries/LibJS/AST.cpp:64
#104740 0x00000000005df79f in JS::update_function_name (value=..., name=...) at ../Libraries/LibJS/AST.cpp:64
#104741 0x00000000005e3087 in JS::AssignmentExpression::execute (this=<optimized out>, interpreter=..., global_object=...) at ../Libraries/LibJS/AST.cpp:1275
#104742 0x00000000005062a2 in JS::Interpreter::execute_statement (this=<optimized out>, global_object=..., statement=..., arguments=..., scope_type=<optimized out>) at ../Libraries/LibJS/Interpreter.cpp:91
#104743 0x00000000005d4d80 in JS::ScopeNode::execute (this=0x2c0, interpreter=..., global_object=...) at ../Libraries/LibJS/AST.cpp:79
#104744 0x0000000000505d48 in JS::Interpreter::run (this=0x1463e20, global_object=..., program=...) at ../Libraries/LibJS/Interpreter.cpp:76
#104745 0x00000000004ebe02 in LLVMFuzzerTestOneInput (data=<optimized out>, size=<optimized out>) at ../Meta/Lagom/Fuzzers/FuzzJs.cpp:70
#104746 0x000000000047cda2 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) ()
#104747 0x000000000047c4e6 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) ()
#104748 0x000000000047e788 in fuzzer::Fuzzer::MutateAndTestOne() ()
#104749 0x000000000047f486 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) ()
#104750 0x000000000046de3f in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) ()
#104751 0x0000000000496c83 in main ()
```

### PR DESCRIPTION
**Pull request by alimpfard**
Fixes #3471, adds a test.

cc @linusg - I think you wrote that `update_function_name()`.

### CODE REVIEW

**Comment by alimpfard** [comment]
Not sure if the test makes sense, I could just put no `expect()` in it, but that didn't seem right either.
I briefly tried adding a `expect().not.toCrash()`, but we can't really recover from this specific case as `malloc` straight-up kills the process.

File: Libraries/LibJS/Tests/functions/function-name.js [hunk 1]
```diff
@@ -48,3 +48,10 @@ test("names of native functions", () => {
     expect((console.debug.name = "warn")).toBe("warn");
     expect(console.debug.name).toBe("debug");
 });
+
+test("cyclic members should not cause infinite recursion (#3471)", () => {
+    let a = [()=>4];
```
Comment (linusg) [inline Libraries/LibJS/Tests/functions/function-name.js:53]: please format tests with prettier :)

*probably:*

```suggestion
    let a = [() => 4];
```

**Review by linusg — COMMENTED** [review]
> cc @linusg - I think you wrote that update_function_name().

I did, I'm still not sure if it's actually part of the ECMAScript spec (we did this based on observable behaviour of other browsers/engines) and I still hate the person who though this is a good idea.

File: Libraries/LibJS/Tests/functions/function-name.js (hunk 1 above)
Comment (alimpfard) [reply Libraries/LibJS/Tests/functions/function-name.js]: Some day I'll remember about this :P

**Pull request merged**